### PR TITLE
move bceWithLogits from python to Aten

### LINKED ISSUE #10648: [feature request] Move binary_cross_entropy_with_logits into ATen
@ebetica mentioned it would be nice to have [`torch.binary_cross_entropy_with_logits`](https://pytorch.org/docs/stable/nn.html#torch.nn.functional.binary_cross_entropy_with_logits) in C++. Seems like a reasonable thing to move into ATen. It's currently implemented in Python.

The only tricky bit is that the the reduction is currently taken as a string `"mean" | "sum" | "none"`. To support all three, we could have the C++ function just return the unreduced tensor, then you can write `torch::binary_cross_entropy_with_logits(...).sum()` or `.mean()` yourself, which is fine. Same in Python, we can just call `.sum()` or `.mean()` on the result in the implementation of `binary_cross_entropy_with_logits` (i.e. no change to the public Python API).

CC @ailzhang

### PR DESCRIPTION
**Pull request by ailzhang**
Fixes #10648 . 
Perf comparison:
```
import torch
import torch.nn as nn
import time

def bm(testsize, repeat=100, cuda=False):
    total_time = 0.0
    pos_weight= torch.ones(testsize[1], device='cuda' if cuda else 'cpu') / testsize[1]
    # loss = nn.BCEWithLogitsLoss(pos_weight=pos_weight)
    loss = nn.BCEWithLogitsLoss()
    input = torch.randn(testsize, device='cuda' if cuda else 'cpu').clamp_(2.8e-2, 1 - 2.8e-2)
    target = torch.randn(testsize, device='cuda' if cuda else 'cpu').gt(0).float()
    input.requires_grad = True
    target.requires_grad = True
    for _ in range(repeat):
        start = time.time()
        l = loss(input, target)
        l.backward()
        # print(target.grad)
        end = time.time()
        total_time += end - start
    return total_time

for cuda in [False, True]:
    for testsize in [(100, 100), (1000, 1000), (2000, 2000)]:
        # print(testsize, cuda)
        print('{:.5f}'.format(bm(testsize, cuda=cuda)))
```
|    | Python CPU | Aten CPU | Python GPU | Aten GPU
| ------------- | ------------- | ------------- | ------------- | ------------- |
| (100, 100)  | 0.15813s | 0.10890s | 0.14601s | 0.07070s |
| (1000, 1000)  | 1.74051s | 0.95038s | 0.15158s | 0.10153s | 
| (2000, 2000) | 5.36515s | 2.46996s | 0.31322s | 0.200941s |

### CODE REVIEW

**Comment by ezyang** [comment]
That thing where c++ is slower than Python... the feeels

**Review by ezyang — APPROVED** [review]
I did not math check the backwards formula.

**Comment by FDecaYed** [comment]
Hi @ailzhang @ezyang,
We(at nvidia) are also aware of the perf issue of this loss function. While moving it to aten get python out of the way, it only solves part of the problem. The other half of this being a bunch of small point-wise kernel launched in GPU case that achieved bad bandwidth overall.
We used to solve this problem as a whole by jitting(and fusing) this from python side. It no longer work now and the new aten implementation does not fuse those ops. What would be the better way to improve this further in your mind?
++ @ngimel and @jjsjann123 for visibility.

A interesting side effect of this change:
the reduce to compute mean/sum used to invoke a DtoH memcpy and cpu side synchronize, which is potentially bad for some workload where CPU resource is valuable. It is no longer the case after this change, but may still exist in other loss function.
